Forms

Timowest12 opened the pull request:
In this PR:

- Added form to add posts to the blog
- Added functionality to add like to post
- Added form to add a comment to post
- Added buttons in UI to link to forms

Timowest12 committed:
Add comment route

Timowest12 committed:
Add comment methods to controller

Timowest12 committed:
Add comment views

Timowest12 committed:
Remove unneeded display

Timowest12 committed:
Add like route

Timowest12 committed:
Add like methods

Timowest12 committed:
Add like buttons

Timowest12 committed:
Add posts controller

Timowest12 committed:
Add posts form

Timowest12 committed:
Add posts routes

Timowest12 committed:
Add posts views

Timowest12 committed:
Fix rubocop errors

SajjadAhmad14 commented on app/controllers/users_controller.rb:
```diff
@@ -7,4 +7,16 @@ def show
+    @newpost = Post.create(user_id: params[:id], title: @postt[:title], text: @postt[:text])
+    @posts_list = @user.three_recent_post
+    render :show
+  end
```
Please shift this code to `PostsController`. A `new ` and `create` actions in `PostsController` can be created. KIndly always try to create `RESTful` actions in the right controller so here the right controller would be `PostsController`, not `usersController`. You may read [this](https://dzone.com/articles/fighting-custom-actions-in-rails-controllers) tutorial for more information on `RESTful` controllers.

Timowest12 replied:
shifted posts methods to posts controller

SajjadAhmad14 commented on app/controllers/users_controller.rb:
```diff
@@ -7,4 +7,16 @@ def show
     @posts_list = @user.three_recent_post
   end
+
+  def new_like
```
Please create a new `RESTful` controller `LikesController` and shift this code to create a new Like there. You can see that this action is to create a new resource of the `Likes` model so it should be done in the right controller.

Timowest12 replied:
Added likes controller and shifted methods to there

SajjadAhmad14 commented on app/controllers/posts_controller.rb:
```diff
@@ -8,4 +8,24 @@ def show
+  def new_comment
+    @user = User.find(params[:user_id])
+    @post = Post.find(params[:id])
+  end
```
Please create a new `RESTful` controller `CommentsController` and shift `new` and `create` actions to create a new Comments resource there. `PostsController` should have only those actions which create posts for a logged-in (Current user). Kindly add a form in the `views/comments/new.html.erb` file which accepts input for creating comments.

Timowest12 replied:
Added comments controller and shifted methods to there

Emmyn5600 replied:
- [ ] Instead of creating the sample comments inside the db/seed, kindly add the form for the comments so that any user can comment depending on the post.

Emmyn5600 replied:
![image](https://user-images.githubusercontent.com/55635977/157853270-bd9cdf3c-ceaf-4c7a-8a02-348b5d1f5982.png)

Emmyn5600 replied:
![image](https://user-images.githubusercontent.com/55635977/157853321-94788aa8-ca59-4db6-aaab-283e6ae5820c.png)

Timowest12 replied:
I did that as well, where the comments are displayed there are 2 links, one says add like and one says add comment. The comments in seed are just to start it up with some data

Emmyn5600 replied:
- [ ] You are setting the comments from the db/seed instead of giving a user an option to add his or her own comment.

Emmyn5600 replied:
@Timowest12 please add the comment’s form in the view and the button or icon for liking the posts. So that a user can see an option to comment or like the post.

SajjadAhmad14 commented on app/controllers/posts_controller.rb:
```diff
@@ -8,4 +8,24 @@ def show
+    @post = Post.find(params[:id])
+    @newlike = Like.create(post: @post, user_id: 1)
+    render :index
+  end
```
Please shift actions to create a `Likes` resource to `LikesController`. PostsController should have only those actions which create posts for a logged-in (Current user).

Timowest12 replied:
Added likes controller and shifted methods to there

Emmyn5600 replied:
- [ ] Please add the button or icon to like the post that has been posted by the user instead of setting everything in the db/seed.

Timowest12 replied:
The button to add comments is on the posts/index page, where all the posts and comments are displayed as well as the likes and likes button

SajjadAhmad14 commented on app/controllers/posts_controller.rb:
```diff
@@ -8,4 +8,24 @@ def show
+    @newlike = Like.create(post: @post, user_id: 1)
+    render :index
+  end
+
```
Please add a private method named `post_params` here to achieve the goal of `strong parameters` in order to protect the model. You may read more about it [here](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/Controllers_lesson.md#:~:text=Pay%20special%20attention%20to%20the%20strong%20parameters%20description%20in%20Rails%20Guides.). Kindly add a private method to `CommentsController` as well to white list the params onve you create `CommentsController`.

Timowest12 replied:
Added private method to make params safe in posts and comments controller

SajjadAhmad14 reviewed (changes requested):
## Required Changes ♻️
> "Change is inevitable. Change is constant. -Benjamin Disraeli"
## Hi Timo, 
 ### You have been doing a tremendous job so far :ok_hand:

## To Highlight 🎉
PR has a title and summary. :heavy_check_mark:
Git flow is followed. :heavy_check_mark:
## To Improve 🙏

- [ ] Please check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### Please remember to tag me in your question so I can receive the notification.
**_Note: Please do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

Timowest12 committed:
Refactor new post function

Timowest12 committed:
Add to refactor

Timowest12 committed:
Refactor likes to correct place

Timowest12 committed:
Refactor like functionality to correct place

Timowest12 committed:
clean up code

Timowest12 committed:
Add Params private methods

Timowest12 committed:
fix rubocop errors

Timowest12 commented:
Changes made:

- Refactored comments methods to be in the correct controller
- Refactored likes methods to be in the correct controller
- Refactored posts methods to be in the correct controller
- Added method to make params private

Emmyn5600 reviewed (changes requested):
Hi @Timowest12 

Good job so far! I know you have worked hard on making this milestone But  There are some issues that you still need to work on to go to the next project but you are almost there!

## To highlight :champagne:

- All linters are passing :heavy_check_mark:
- Good commit history :heavy_check_mark:
- Good PR description :heavy_check_mark:

## Required Changes :recycle:

- [ ] Check the comment under the review

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!:clap::clap::clap:

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
#### **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Timowest12 commented:
hallo @Emmyn5600 
In the comments you added you say all of my comments are added through seeds, this is not the case, on the page that displays all posts and comments there are 2 buttons one to add a like and one to add a comment(see screenshot)

![Screenshot from 2022-03-11 12-37-59](https://user-images.githubusercontent.com/13661892/157861108-0743fabd-e80d-4638-b033-a33b054b133c.png)

Since the wireframe does not display where exactly the forms should be displayed I just placed it where thought would be the most logical.

the comments that are in the seed are only there to have some data when starting up the program.
both functions work and the previous reviewer only required changes to place the controller in another file.
please tell me if you want me to still change anything.
Greets Timo.

Emmyn5600 commented:
@Timowest12 I can't see those buttons on my end.

Emmyn5600 reviewed (approved):
Hi @Timowest12 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! :tada: :clap: :clap: :100:

![image](https://user-images.githubusercontent.com/55635977/157864238-af7914f2-f381-4f64-b9de-74be213e07c4.png)

![nodding-yes](https://user-images.githubusercontent.com/65068771/143095390-344d1cb0-b853-4b91-8662-5d9049dc9e60.gif)

## Optional suggestions

- [ ] N/A

#### _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!:clap::clap::clap:
#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
------
#### _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).
